alf/environments/metadrive/sensors.py

```diff
@@ -0,0 +1,158 @@
+from alf.tensor_specs import TensorSpec
+from .geometry import FieldOfView, Polyline
+from .map_perception import MapPolylinePerception
+
+
+@alf.configurable
+class VectorizedObservation(ObservationBase):
+    """This implements a customized observation for the MetaDrive environment that
+    produces vectorized inputs (as opposed to raster inputs such as BEV).
+
+    The main API observe() is designed to produces a dictionary of vectorized
+    inputs. Depending on the configuration the dictionary may consist of:
+
+    1. 'map': polyline based features of road network (lanes) and navigations
+    2. 'ego': polyline based feature of ego car history
+    3. 'agents': polyline based feature of dynamic road users for interaction
+
+    All polylines are within a specified field of view, and transformed so that
+    they are in the ego car's body frame, where x-axis points to the heading
+    direction of the ego car.
+
+    """
+
+    def __init__(self,
+                 vehicle_config: metadrive.utils.Config,
+                 fov: FieldOfView = FieldOfView(),
+                 segment_resolution: float = 2.0,
+                 polyline_size: int = 4,
+                 polyline_limit: int = 64,
+                 history_window_size: int = 8):
+        """Construct a VectorizedObservation instance.
+
+        Args:
+
+            vehicle_config: A MetaDrive global configuration of the MetaDrive
+                environment that this observation is attached to.
+            fov: Defines the field of view with respect to the ego car. Map object
+                and agents outside of the field of view will not appear in the
+                observation. Usually the bigger the field of view, the more expensive
+                the computation of the obsrevation (and the training) will be.
+            segment_resolution: The length of each line segment in the polylines
+                during sampling. The smaller the value, the more segments and
+                polylines. As a result, it also implies more expensive training.
+            polyline_size: Specify the number of segments in one polyline. Putting
+                more segments in a polyline can reduce the number of polylines for
+                each observation.
+            polyline_limit: Specify the maximum number of polylines in the
+                observation for map and navigation. If the actual number of polylines
+                goes beyond this limit, farthest polylines (from the ego car) will be
+                filtered out until the limit is satisfied. If the actual number of
+                polylines is below this limit, zero padding will be employed to bring
+                fill the vacancies.
+            history_window_size: The past positions of the most recent this number of
+                frames will be recorded and used for the ego history feature.
+
+        """
+        super().__init__(vehicle_config)
+        self._map_perception = MapPolylinePerception(
```

Comment:
what is the rationale of having a separate MapPolylinePerception class? It seems that VectorizedObservation is a very thin wrapper on top of MapPolylinePerception. I feel these two can be merged.

Reply:
If there were only MapPerception, yes it should be merged (and that is actually how this class look like before agent perception are added in, i.e., there used to be only `VectorizedObervation` class).

With the `AgentPerception` added, I felt it is more clear to have split the code into 2 classes (hence `MapPolylinePerception` and `AgentPerception`, where the latter is introduced in the [next PR](https://github.com/HorizonRobotics/alf/pull/1181)) because both of them is non-trivial. The ego perception, however, is trivial enough so that it does not deserve its own modular design.

Reply:
I guess this is a problem of separating a change into too many parts. Although a giant PR is not desirable, the other extreme is not good either.

Reply:
Yes, apparently everything comes at a cost. Sorry that my intention to make it easier for the reviewer did not achieve the goal here. Will try to strike a better trade off netxt time.